src/device-manager/java/src/nl/Weave/DataManagement/GenericTraitUpdatableDataSink.java

```diff
@@ -0,0 +1,78 @@
+/*
+
+    Copyright (c) 2019 Google, LLC.
+    All rights reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Represents GenericTraitUpdatableDataSink Interface.
+ */
+
+package nl.Weave.DataManagement;
+
+import android.os.Build;
+import android.util.Log;
+
+import java.math.BigInteger;
+import java.util.EnumSet;
+import java.util.Random;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Iterator;
+
+public interface GenericTraitUpdatableDataSink
+{
+    public void setSigned(String path, int value, boolean isConditional);
```

Comment:
For setSigned() methods:

```
/**
 * Assigns the provided value to the given path as a signed integer value.
 * 
 * @param path the proto path to the property to modify
 * @param value the new value to assign to the property
 * @param isConditional whether or not to allow overwriting any conflicting changes. If true, then if a later
 *     version of the trait has modified this property, this update will be dropped; otherwise, this value will 
 *     overwrite the newer change
 */
```
 
For setUnsigned() methods:

```
/**
 * Assigns the provided value to the given path as an unsigned integer value.
 * 
 * @param path the proto path to the property to modify
 * @param value the new value to assign to the property
 * @param isConditional whether or not to allow overwriting any conflicting changes. If true, then if a later
 *     version of the trait has modified this property, this update will be dropped; otherwise, this value will 
 *     overwrite the newer change
 */
```
 
For general set() methods:

```
/**
 * Assigns the provided value to the given path.
 * 
 * @param path the proto path to the property to modify
 * @param value the new value to assign to the property
 * @param isConditional whether or not to allow overwriting any conflicting changes. If true, then if a later
 *     version of the trait has modified this property, this update will be dropped; otherwise, this value will 
 *     overwrite the newer change
 */
```
 
For getters() could do something like this, although throwing a generic "Exception" isn't great. Not sure what the C++ code throws for this, but it could be nice to define our own exceptions like `InvalidPathException` or `InvalidTypeException` and throw specific errors depending on what case is causing the failure.

```
/**
 * Returns the value assigned to the property at the given path within this trait.
 *
 * @throws Exception if no property exists at this path, or if the type of the property does not match
 */
```
 
For getVersion()

```
/** Returns the version of the trait represented by this data sink. */
```

For beginRefreshData():

```
/** 
 * Begins a sync of the trait data. The result of this operation can be observed through the {@link CompletionHandler}
 * that has been assigned via {@link #setCompletionHandler}.
 */
```

Reply:
thanks, will add the rest of them.

Reply:
regarding the specific exception, InvalidPathException , InvalidTypeException , it looks fine, I would prefer address it in later PR on it